slack_driver/driver.py

```diff
@@ -0,0 +1,186 @@
+            if e.get('primary'):
+                user_email = e.get('value')
+                logger.debug('Found primary email {}'.format(user_email))
+                break
+
+        # Just in case Slack API somehow doesn't have a primary email. This should normally not happen.
+        if user_email is None:
+            user_email = user.get('emails')[0].get('value')
+            logger.warning('No slack primary email set for user {}, '
+                           'selecting first email in list {}'.format(e.get('id'), user_email))
+
+        if user.get('active'):
+            users[user_email] = user.get('id')
+        else:
+            users_already_deactivated[user_email] = user.get('id')
+
+    failure = 0
+    users_to_enable = set(users_already_deactivated.keys()) & set(allowed_users.keys())
+    logger.debug('Will re-enabled {} user(s) which were deactivated, '
+                 'but got access to Slack again'.format(len(users_to_enable)))
+
+    for u in users_to_enable:
+        logger.debug('Will now enable {} ({})'.format(u, users_already_deactivated[u]))
+        try:
+            ret = sc.activate_user(users_already_deactivated[u])
+            if (ret.get('active') is not True):
+                logger.warning('Failed to enable user {}'.format(users_already_deactivated[u]))
+                failure = failure + 1
+        except Exception as e:
+            logger.warning('Could not enable user {}, exception: {}'.format(users_already_deactivated[u], e))
+            failure = failure + 1
+
+    users_to_disable = set(users.keys()) - set(allowed_users.keys())
+    logger.debug('Will disable {} user(s) which should no longer have access to Slack'.format(len(users_to_disable)))
+
+    for u in users_to_disable:
+        logger.debug('Will now disable user {} ({})'.format(u, users[u]))
+        # sample return msg:
+        # {"schemas": ["urn:scim:schemas:core:1.0"], "id": "UB0GWPDCM", "externalId": "", "meta": {"created":
+        # "2018-06-01T16:10:18-07:00", "location": "https://api.slack.com/scim/v1/Users/UB0GWPDCM"}, "userName":
+        # "kang_slack", "nickName": "kang_slack", "name": {"givenName": "", "familyName": ""}, "displayName": "",
+        # "profileUrl": "https://mozilla-sandbox-scim.slack.com/team/kang_slack", "title": "", "timezone":
+        # "America/Los_Angeles", "active": false, "emails": [{"value": "kang+slack@mozilla.com", "primary": true}],
+        # "photos": [{"value":
+        # "https://example.net/test.png",
+        # "type": "photo"}], "groups": []}
+
+        try:
+            ret = sc.deactivate_user(users[u])
+            if (ret.get('active') is not False):
+                logger.warning('Failed to disable user {}'.format(users[u]))
+                failure = failure + 1
+        except Exception as e:
+            # This can happen if the user is unmodifiable, for example if its a space owner
+            logger.warning('Could not disable user {}, exception: {}'.format(users[u], e))
```

Comment:
Is there some better way to deal with these cases?  Like set a service account to space owner and then disable anyway?

Reply:
the api cannot do this afaik - but we should definitely set things up so that a service account is owner, not a user (so that this never happens, and its much safer anyway regardless of how the driver operates)